test/test_images/listener/main.go

```diff
@@ -32,18 +44,29 @@ func main() {
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	// Launch a go routine to avoid blocking
-	go func() {
-		timeout := 10 * time.Second
-		<-time.After(timeout)
-		logging.FromContext(ctx).Infow("cancelling context: timeout reached", "timeout", timeout)
-		cancel()
-	}()
+	var env envConfig
+	if err := envconfig.Process("", &env); err != nil {
+		log.Fatalf("Unable to read environment config: %v", err)
+	}
+
+	numExpectedEvents, err := strconv.Atoi(env.ExpectedEventCount)
+	if err != nil {
+		log.Fatalf("Invalid value for EVENT_COUNT (%v): %v", env.ExpectedEventCount, err)
+	}
 
 	var count int32
 	if err = client.StartReceiver(ctx, func(ctx context.Context, event cloudevents.Event) {
 		logging.FromContext(ctx).Infof("Received event: %s", event.String())
-		atomic.AddInt32(&count, 1)
+
+		if event.Type() == env.ExpectedEventType {
+			atomic.AddInt32(&count, 1)
```

Comment:
Instead of using `atomic`, I tend to collect the messages in a slice using channels, i.e.:

```go
events := make(chan cloudevents.Event, numExpectedEvents)
go func() {
  if err = client.StartReceiver(ctx, func(...) {
    if event.Type() == env.ExpectedEventType {
      events <- event
    }
    // Do rest of processing outside handler
  }); err != nil {
    // whinge and exit
  }
}()

for event := <-events {
  // Process events one by one, keeping count. When complete, call `cancel()` to exit StartReceiver, and break to end the loop.
}
```

Your way works as well, except for your cast from `int` to `int32` on line 79. One advantage to the channel approach is that you can handle each message serially and don't have to worry about mutexing any shared data structures, so you can handle the loop more "obviously". I'd also consider extracting the CloudEvent handler function so that you don't get a 10-line anonymous goroutine for the `go func() {...}()` construct.